js/apps/admin-ui/src/i18n/OverridesBackend.ts

```diff
@@ -0,0 +1,160 @@
+      return data;
+    }
+
+    // Ensure we are operating on a cloned data structure to prevent in place mutations.
+    const target = structuredClone(data);
+
+    for (const [path, value] of Object.entries(overrides[namespace])) {
+      this.#applyOverride(target, path, value);
+    }
+
+    return target;
+  }
+
+  /** Applies an override by converting path segments denoted with a key separator as nested objects and merging the result. */
+  #applyOverride(target: Record<string, any>, path: string, value: string) {
+    const trail = path.split(KEY_SEPARATOR);
+    let pointer = target;
+
+    trail.forEach((segment, index) => {
+      const isLast = index === trail.length - 1;
+      pointer = pointer[segment] = isLast ? value : pointer[segment] ?? {};
+    });
+  }
+
+  #loadOverrides(languages?: string | string[]) {
+    const locale = this.#determineLocale(languages);
+    const cachedOverrides = this.#overridesCache.get(locale);
+
+    if (cachedOverrides) {
+      return cachedOverrides;
+    }
+
+    const overrides = adminClient.realms
+      .getRealmLocalizationTexts({
+        realm: adminClient.realmName,
+        selectedLocale: locale,
+      })
+      .then((data) => this.#parseOverrides(data));
+
+    this.#overridesCache.set(locale, overrides);
+
+    // Evict cached request on failure.
+    overrides.catch((error) => {
+      this.#overridesCache.delete(locale);
+      return Promise.reject(error);
+    });
+
+    return overrides;
+  }
+
+  #parseOverrides(data: Record<string, string>) {
+    const parsed: ParsedOverrides = {};
+
+    for (const [path, value] of Object.entries(data)) {
+      const parts = path.split(NAMESPACE_SEPARATOR);
+      const namespace = parts.length > 1 ? parts[0] : DEFAULT_NAMESPACE;
+      const key = parts.length > 1 ? parts[1] : parts[0];
+
+      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+      if (!parsed[namespace]) {
+        parsed[namespace] = {};
+      }
+
+      parsed[namespace][key] = value;
+    }
+
+    return parsed;
+  }
+
+  #determineLocale(languages?: string | string[]) {
+    if (typeof languages === "string") {
+      return languages;
+    }
+
+    return languages?.[0] ?? DEFAULT_LOCALE;
+  }
+
+  #determineNamespace(namespaces?: string | string[]) {
+    if (typeof namespaces === "string") {
+      return namespaces;
+    }
+
+    return namespaces?.[0] ?? DEFAULT_NAMESPACE;
+  }
+
+  #loadUrlPromisified(
+    url: string,
+    languages?: string | string[],
+    namespaces?: string | string[]
+  ) {
+    return new Promise<ResourceKey>((resolve, reject) => {
+      const callback: ReadCallback = (error, data) => {
+        if (error) {
+          return reject(error);
+        }
+
+        if (
+          typeof data !== "object" ||
+          data === null ||
+          typeof data === "boolean"
+        ) {
+          return reject(
+            new Error(
+              "Unable to load URL, data returned is of an unsupported type."
```

Comment:
What about also providing the `url` as part of the error message?

Reply:
I'll add the [`cause`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause) option, this should allow debuggers to show the original error that caused this issue.